score.go

```diff
@@ -0,0 +1,952 @@
+		return fmt.Errorf("invallid FirstMessageDeliveriesWeight; must be positive (or 0 to disable)")
+	}
+	if p.FirstMessageDeliveriesWeight != 0 && (p.FirstMessageDeliveriesDecay <= 0 || p.FirstMessageDeliveriesDecay >= 1) {
+		return fmt.Errorf("invalid FirstMessageDeliveriesDecay; must be between 0 and 1")
+	}
+	if p.FirstMessageDeliveriesWeight != 0 && p.FirstMessageDeliveriesCap <= 0 {
+		return fmt.Errorf("invalid FirstMessageDeliveriesCap; must be positive")
+	}
+
+	// check P3
+	if p.MeshMessageDeliveriesWeight > 0 {
+		return fmt.Errorf("invalid MeshMessageDeliveriesWeight; must be negative (or 0 to disable)")
+	}
+	if p.MeshMessageDeliveriesWeight != 0 && (p.MeshMessageDeliveriesDecay <= 0 || p.MeshMessageDeliveriesDecay >= 1) {
+		return fmt.Errorf("invalid MeshMessageDeliveriesDecay; must be between 0 and 1")
+	}
+	if p.MeshMessageDeliveriesWeight != 0 && p.MeshMessageDeliveriesCap <= 0 {
+		return fmt.Errorf("invalid MeshMessageDeliveriesCap; must be positive")
+	}
+	if p.MeshMessageDeliveriesWeight != 0 && p.MeshMessageDeliveriesThreshold <= 0 {
+		return fmt.Errorf("invalid MeshMessageDeliveriesThreshold; must be positive")
+	}
+	if p.MeshMessageDeliveriesWindow < 0 {
+		return fmt.Errorf("invalid MeshMessageDeliveriesWindow; must be non-negative")
+	}
+	if p.MeshMessageDeliveriesWeight != 0 && p.MeshMessageDeliveriesActivation < time.Second {
+		return fmt.Errorf("invalid MeshMessageDeliveriesActivation; must be at least 1s")
+	}
+
+	// check P3b
+	if p.MeshFailurePenaltyWeight > 0 {
+		return fmt.Errorf("invalid MeshFailurePenaltyWeight; must be negative (or 0 to disable)")
+	}
+	if p.MeshFailurePenaltyWeight != 0 && (p.MeshFailurePenaltyDecay <= 0 || p.MeshFailurePenaltyDecay >= 1) {
+		return fmt.Errorf("invalid MeshFailurePenaltyDecay; must be between 0 and 1")
+	}
+
+	// check P4
+	if p.InvalidMessageDeliveriesWeight >= 0 {
+		return fmt.Errorf("invalid InvalidMessageDeliveriesWeight; must be negative")
+	}
+	if p.InvalidMessageDeliveriesDecay <= 0 || p.InvalidMessageDeliveriesDecay >= 1 {
+		return fmt.Errorf("invalid InvalidMessageDeliveriesDecay; must be between 0 and 1")
+	}
+
+	return nil
+}
+
+// router interface
+func (ps *peerScore) Start(gs *GossipSubRouter) {
+	if ps == nil {
+		return
+	}
+
+	ps.msgID = gs.p.msgID
+	ps.host = gs.p.host
+	go ps.background(gs.p.ctx)
+}
+
+func (ps *peerScore) Score(p peer.ID) float64 {
+	if ps == nil {
+		return 0
+	}
+
+	ps.Lock()
+	defer ps.Unlock()
+
+	return ps.score(p)
+}
+
+func (ps *peerScore) score(p peer.ID) float64 {
+	pstats, ok := ps.peerStats[p]
+	if !ok {
+		return 0
+	}
+
+	var score float64
+
+	// topic scores
+	for topic, tstats := range pstats.topics {
+		// the topic score
+		var topicScore float64
+
+		// the topic parameters
+		topicParams, ok := ps.params.Topics[topic]
+		if !ok {
+			// we are not scoring this topic
+			continue
+		}
+
+		// P1: time in Mesh
+		if tstats.inMesh {
+			p1 := float64(tstats.meshTime / topicParams.TimeInMeshQuantum)
```

Comment:
We should globally cap the sum total as well. Else, an attacker could boost their score artificially by meshing with us on many topics, and staying on those for a long time.

Reply:
We already have an implicit global cap: each individual topic score is capped, so the global cap is the weight sum of the caps.

Reply:
I was thinking of a scenario where an attacker could enumerate all topics we're interested in, and GRAFT with us on all of them, whereas the average honest peer only GRAFTs with us on a subset of topics. That attacker could accumulate a higher score by the mere fact that they're behaving well on more topics than the rest. But I guess that's how the algorithm is designed to work? Also, the probability of us accepting all of their `GRAFT`s would be low.

Reply:
On further thought it would make sense to cap the topic score contribution; I'll add an (optional) total topic score cap parameter.

Reply:
Implemented an (optional) TopicScoreCap parameter that produces the desired effect.